[Feat] #4 WeatherService와 Stub 구현

jiyoungsophia opened the pull request:
<!--
  🙌 풀 리퀘스트 제목은 아래와 같이 해주세요!
      <종류>: <이슈 번호> <제목>
      ex: [Feat] #167 예약 취소 구현
  ✔️ Optional, 담당자 (자신), 라벨 설정했는지 확인하세요
-->
## About this PR
### ⚓ Related Issue
<!-- 관련된 이슈 번호를 적어주세요. -->
- #4

<br>

### 🦁 Contents
<!-- 이 PR에서 작업한 내용에 대해 알려주세요! -->

- **WeatherDTO.swift**: OpenWeatherMap API 4개 엔드포인트별 Response DTO 정의
- **CurrentWeather+.swift**: 4개 API 응답을 하나의 Domain Model로 변환하는 수동 디코딩
  - 4개의 API를 조합하고, 응답구조도 복잡하다고 판단해 Mapper 대신 생성자 방식 선택했습니다
- **WeatherAPI.swift**: Moya TargetType으로 4개 API 엔드포인트 정의, JSON stub 데이터 연결
- **WeatherService.swift**: RxMoya로 4개 API 병렬 호출 후 zip으로 결합하는 서비스 구현
    - getCurrentWeather함수의 매개변수로 (lat: Double, lon: Double)과 CLLocationCoordinate2D을 고민하다가 타입 안정성을 위해 후자를 선택했습니다
- WeatherServiceStub.swift: JSON 파일 기반 stub 구현
- DateFormatter+.swift: 한국어 시간/요일 포맷 변환 함수들
- Bundle+.swift: Info.plist에서 API 키 읽어오는 extension
- HourlyForecast.swift: hour 타입을 Int → String으로 변경
- 4개 JSON 파일: 실제 API 응답과 동일한 구조의 테스트 데이터
- Info.plist: API 키 환경변수 설정 추가


<br>

## Other information 🔥
<!-- 다른 리뷰어가 참고하면 좋을 내용을 알려주세요. 기타 참고사항이 있다면 작성해줍니다. -->

**Stub 사용법 (API 키가 필요하지 않습니당)**

```swift
// Reactor에서 WeatherService 사용
final class WeatherReactor: Reactor {
   private let weatherService: WeatherServiceType
   
   init() {
       // Stub 데이터로 ~!~!!
       self.weatherService = WeatherService(provider: WeatherService.stubProvider)
     }
      ...
   }
   ```
   
   **그치만 Config 파일 설정 필요,,!!**
- 프로젝트 루트에 `Config.xcconfig` 파일 생성 후 다음 내용 추가:
```
WEATHER_API_KEY = {api_key}
```

jiyoungsophia committed:
chore: #4 - openweather api 키 설정

jiyoungsophia committed:
feat: #4 - Moya 사용을 위한 WeatherAPI TargetType 구현

jiyoungsophia committed:
Merge branch 'develop' into feat/#4-networkservice

# Conflicts:
#	SajaWeather/SajaWeather.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
#	SajaWeather/SajaWeather/Info.plist

jiyoungsophia committed:
feat: #4 - 4개의 API에 대한 응답 DTO를 CurrentWeather 뷰 모델로 변환하는 Mapper 구현

jiyoungsophia committed:
feat: #4 - RxMoya 사용해 날씨 정보 요청하는 WeatherService 구현

jiyoungsophia committed:
chore: #4 - WeatherService용 stub 데이터 및 stubprovider 추가

Jaesung-Jung commented on SajaWeather/SajaWeather/Data/Service/Network/WeatherService.swift:
```diff
@@ -0,0 +1,75 @@
+        coordinate: coordinate
+      )
+    }
+    .catch { error in
+      return Observable.error(error)
+    }
```
이 연산은 없어도 될것 같아요. catch 에서 error Observable을 만드는 것은 없어도 완전 동일합니다.

jiyoungsophia replied:
아!!! 그냥 똑같은 에러로 다시 방출하는거라 의미가 없었군요,,!!! 감사합니다!!!!

housedone commented on SajaWeather/SajaWeather/Data/Model/HourlyForecast.swift:
```diff
@@ -8,7 +8,7 @@
 
 struct HourlyForecast {
-  let hour: Int
+  let hour: String
```
아 이 부분 셀 짤때 String으로 했는데 맞춰주셨네요 감사합니다!

housedone commented on SajaWeather/SajaWeather/Utils/Extensions/DateFormatter+.swift:
제가 생각하는 게 맞는지 자신은 없습니다만 ㅋㅋ;
- 시간 포맷 부분에서 timeFormatter의 요 부분은 한번 확인해주시면 될 것 같습니다.
  - 24시간제라면 h가 아닌 H인거 같은데 이게 저희 일출 일몰 나타낼 때 쓰는 부분인거죠? 
  - 그렇다면 7:35PM 같은 식이니 h가 맞는 것 같고
  - MM은 2자리 월 표기이고 mm이 2자리 분을 나타내는 것 같습니다.

- static 함수들이 호출될 때마다 새로운 DateFormatter를 생성하는데, 얘들을 딱 static let으로 선언해서 재사용한다면 어떨까요?

``` swift
import Foundation

extension DateFormatter {
    /// 12AM, 1PM
    private static let hourlyFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "ha"
        formatter.locale = Locale(identifier: "en_US")
        return formatter
    }()

    /// 17:29
    private static let timeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "h:mm"
        formatter.locale = Locale(identifier: "en_US")
        return formatter
    }()

    /// 당일이면 "오늘", 나머지는 "8월 8일 금요일" 반환
    private static let dayFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "MM월 dd일 EEEE"
        formatter.locale = Locale(identifier: "ko_KR")
        return formatter
    }()

    static func formatHour(_ timeInterval: TimeInterval) -> String {
        let date = Date(timeIntervalSince1970: timeInterval)
        return hourlyFormatter.string(from: date)
    }

    static func formatTime(_ timeInterval: TimeInterval) -> String {
        let date = Date(timeIntervalSince1970: timeInterval)
        return timeFormatter.string(from: date)
    }

    static func formatDay(_ timeInterval: TimeInterval) -> String {
        let date = Date(timeIntervalSince1970: timeInterval)
        if Calendar.current.isDateInToday(date) {
            return "오늘"
        }
        return dayFormatter.string(from: date)
    }
}
```

이렇게 한다면 CurrentWeather+ 익스텐션에서 호출 시에 `.timeFormatter` 등을 `.formatTime` 등으로 바꾸면 될 것 같습니다.

jiyoungsophia replied:
DateFormatter 연산이 비싸다는 사실을 알고는 있었지만 정작 어떻게 해결할지는 제대로 생각하지 않았던 게으름,, 반성합니다..!!!
static let 인스턴스가 한번만 만들어지고 그 이후엔 재사용되는 걸 완전 까먹고 있었어요!!! 매번 새로 생성하는 게 아니라 한 번 만들어놓고 계속 쓰는 거라는거!!!!
너무 감사합니다👍 바로 수정할게요!!

Jaesung-Jung replied:
[Date.FormatStyle](https://developer.apple.com/documentation/foundation/date/formatstyle)
이것도 고려해보세요. DateFormatter, NumberFormatter 등을 대체하는 API이고 동일 Format에 대해 자동으로 캐시도 합니다.

housedone commented on SajaWeather/SajaWeather/Data/Model/CurrentWeather+.swift:
하드코딩된 기본값들을 static let으로 나중에 뺀다면 좋을 것 같습니다! 저도 해야 될 것 같고요 ㅋㅋ;

jiyoungsophia replied:
하드 코딩된 기본값들은 어떤걸 의미하는 걸까요?!

housedone replied:
weather.id 에 800, aqi에 1 이런 것들이요!

housedone reviewed (approved):
잘 봤습니다! 
- Moya 잘 써주셔서 너무 감사합니다 
- 역할 분리를 잘 해주신 것 같습니다
- 네트워크 없이도 테스트하기 좋은 여러 stub 추가 감사드립니다.
- 여러 API 요청을 묶어서 openweathermap의 유료 One Call API급으로 만들어주신 부분 정말 감사드립니다.


제 역량 안에서 지영님의 PR에 성심성의껏 답하고자 했지만 부족하거나 제가 잘못 이해한 점 있다면 멱살잡고 마구 알려주세요. 
클린아키텍처 좀 봤는데 저는 아키텍처 꺠우치는 데에 좀 오래 걸리는 것 같습니다.. 
주말 마무리 잘 되시고 계속 화이팅입니다!

jiyoungsophia committed:
chore: #4 - 코드리뷰 반영

- WeatherAPI: 무의미하게 던지던 에러 구문 삭제
- DateFormatter+, WeatherService: DateFormatter 재사용으로 성능 개선 및 메서드명 정리
- 그 외 컨벤션 및 코드 정리

Pjh01 commented on SajaWeather/SajaWeather/Data/Service/Network/WeatherService.swift:
```diff
@@ -0,0 +1,71 @@
+    )
+  }
+  
+  func getCurrentWeather(coordinate: CLLocationCoordinate2D) -> Observable<CurrentWeather> {
```
`Single` 말고 `Observable`을 사용하신 이유가 궁금합니다!!

jiyoungsophia replied:
rxMoya request도 single이 기본 반환값이고, 날씨 데이터 하나만,, 한번만 호출하면 되는데 전형적인 Single인데 왜 observable로 굳이 변환하면서까지 했었는지 모르겠습니다..!!!! 말씀해주셔서 감사합니다!!!

Pjh01 commented on SajaWeather/SajaWeather/Data/Service/Network/WeatherService.swift:
```diff
@@ -0,0 +1,71 @@
+    let lat = coordinate.latitude
+    let lon = coordinate.longitude
+    
+    let currentWeatherObservable = provider.rx
+      .request(.current(lat: lat, lon: lon))
+      .map(CurrentWeatherResponseDTO.self)
+      .asObservable()
+    
+    let hourlyForecastObservable = provider.rx
+      .request(.hourlyForecast(lat: lat, lon: lon))
+      .map(HourlyForecastResponseDTO.self)
+      .asObservable()
+    
+    let dailyForecastObservable = provider.rx
+      .request(.dailyForecast(lat: lat, lon: lon))
+      .map(DailyForecastResponseDTO.self)
+      .asObservable()
+    
+    let airQualityObservable = provider.rx
+      .request(.airQuality(lat: lat, lon: lon))
+      .map(AirQualityResponseDTO.self)
+      .asObservable()
```
비슷한 로직이라 이런 식으로 해도 좋을 것 같아요!
```swift
private func request<T: Decodable>(_ target: WeatherAPI, type: T.Type) -> Observable<T> {
    provider.rx.request(target)
        .map(T.self)
        .asObservable()
}

let currentWeatherObservable = request(.current(lat: lat, lon: lon), type: CurrentWeatherResponseDTO.self)
let hourlyForecastObservable = request(.hourlyForecast(lat: lat, lon: lon), type: HourlyForecastResponseDTO.self)
...
```

jiyoungsophia replied:
공통로직이 단순하고 변경될 일이 없을것 같아 작성했던 코드인데요, 추후에 로직이 복잡해지거나 엔드포인트가 많아지게 되면 제안해주신 방식으로 수정하겠습니다!!

Pjh01 reviewed (approved):
감사합니다~ 잘 사용해볼게요!!

jiyoungsophia committed:
chore: #4 - 코드리뷰 반영

- WeatherService: Single로 반환값 변경
